dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1540 @@
+        {
+            m_popup.set(winrt::Popup());
+            m_popup.get().Child(*this);
+            m_popup.get().Closed({ this, &TeachingTip::OnPopupClosed });
+            if (IsLightDismissEnabled())
+            {
+                m_popup.get().IsLightDismissEnabled(true);
+            }
+        }
+        if (!m_contractAnimation)
+        {
+            CreateContractAnimation();
+        }        
+        if (!m_expandAnimation)
+        {
+            CreateExpandAnimation();
+        }
+
+        if (!m_popup.get().IsOpen())
+        {
+            //We are about to begin the process of trying to open the teaching tip, so notify that we are no longer idle.
+            if (m_isIdle)
+            {
+                m_isIdle = false;
+                TeachingTipTestHooks::NotifyIdleStatusChanged(*this);
+            }
+            m_popup.get().IsOpen(true);
+            if (m_rootGrid)
+            {
+                StartExpandAnimation();
+            }
+            else
+            {
+                m_startAnimationInOnApplyTemplate = true;
+            }
+        }
+        else
+        {
+            //We have become Open but our popup was already open. This can happen when a close is canceled by the closing event, so make sure the idle status is correct.
+            if (!m_isIdle && !m_isExpandAnimationPlaying && !m_isContractAnimationPlaying)
+            {
+                m_isIdle = true;
+                TeachingTipTestHooks::NotifyIdleStatusChanged(*this);
+            }
+        }
+    }
+    else
+    {
+        if (m_popup)
+        {
+            if (m_popup.get().IsOpen())
+            {
+                //We are about to begin the process of trying to close the teaching tip, so notify that we are no longer idle.
+                if (m_isIdle)
+                {
+                    m_isIdle = false;
+                    TeachingTipTestHooks::NotifyIdleStatusChanged(*this);
+                }
+                RaiseClosingEvent(m_lastCloseReason);
+            }
+            else
+            {
+                //We have become not Open but our popup was already not open. Lets make sure the idle status is correct.
+                if (!m_isIdle && !m_isExpandAnimationPlaying && !m_isContractAnimationPlaying)
+                {
+                    m_isIdle = true;
+                    TeachingTipTestHooks::NotifyIdleStatusChanged(*this);
+                }
+            }
+        }
+
+        m_currentEffectivePlacementMode = winrt::TeachingTipPlacementMode::Auto;
+        TeachingTipTestHooks::NotifyEffectivePlacementChanged(*this);
+    }
+    TeachingTipTestHooks::NotifyOpenedStatusChanged(*this);
+}
+
+void TeachingTip::OnIconSourceChanged()
+{
+    if (auto source = IconSource())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Icon"sv, false);
+        if (m_iconContentPresenter)
+        {
+            m_iconContentPresenter.get().Content(SharedHelpers::MakeIconElementFrom(source));
```

Comment:
What prevents this from being in VSM?

Reply:
We don't have a "IconSourcePresenter" that we can put the icon source in. There is not a way (I think?) of getting a vsm object to make an icon element from a source as we do here. We should prehaps make an "IconSourcePresenter" that can do just this.

---
In reply to: [239325312](https://github.com/Microsoft/microsoft-ui-xaml/pull/32#discussion_r239325312) [](ancestors = 239325312)

Reply:
We do have IconSourceElement which would enable this. Unfortunately, it is not available all the way down-level. So I think what you have here is the best that can be done.